wrappers/glmemshadow.cpp

```diff
@@ -0,0 +1,404 @@
+/*
+ * Copyright © 2019 Intel Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "glmemshadow.hpp"
+
+#include <unordered_map>
+#include <algorithm>
+
+#include <assert.h>
+
+#ifdef _WIN32
+
+#include <windows.h>
+
+#else
+
+#include <unistd.h>
+#include <signal.h>
+#include <sys/mman.h>
+
+#endif
+
+#include "gltrace.hpp"
+#include "os_thread.hpp"
+#include "os.hpp"
+
+static bool sInitialized = false;
+
+static std::unordered_map<size_t, GLMemoryShadow*> sPages;
+static size_t sPageSize;
+
+static os::mutex mutex;
+
+enum class MemProtection {
+#ifdef _WIN32
+    NO_ACCESS = PAGE_NOACCESS,
+    READ_ONLY = PAGE_READONLY,
+    READ_WRITE = PAGE_READWRITE,
+#else
+    NO_ACCESS = PROT_NONE,
+    READ_ONLY = PROT_READ,
+    READ_WRITE = PROT_READ | PROT_WRITE,
+#endif
+};
+
+size_t getSystemPageSize() {
+#ifdef _WIN32
+    SYSTEM_INFO info;
+    GetSystemInfo(&info);
+    return info.dwPageSize;
+#else
+    return sysconf(_SC_PAGESIZE);
+#endif
+}
+
+void memProtect(void *addr, size_t size, MemProtection protection) {
+#ifdef _WIN32
+    DWORD flOldProtect;
+    BOOL bRet = VirtualProtect(addr, size, static_cast<DWORD>(protection), &flOldProtect);
+    if (!bRet) {
+        DWORD dwLastError = GetLastError();
+        os::log("apitrace: error: VirtualProtect failed with error 0x%lx\n", dwLastError);
+        exit(EXIT_FAILURE);
```

Comment:
`exit()` might still call C++ destructurors.  It's better on wrapper code to always call `os::abort()`

Reply:
Didn't think about this, thanks.